File: Api/Users/iCompanyPortal.Api.Users.Client/UsersClient.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace iCompanyPortal.Api.Users.Client
{
    public class UsersClient : IUsersClient
    {
        private readonly HttpClient http;

        public UsersClient(HttpClient http)
        {
            this.http = http;
        }

        public async Task<bool> DoesUserExistAsync(int userId)
        {
            return await http.GetAsync<bool>($"/{userId}/exists");
        }

        public async Task<bool> IsEmailUniqueAsync(string email)
        {
            return await http.GetAsync<bool>($"/{WebUtility.UrlEncode(email)}/is-unique");
        }

        public async Task<UserInfo> GetSignedInUserAsync()
        {
            return await http.GetAsync<UserInfo>("/");
        }

        public async Task<UserInfo> GetUserAsync(int userId)
        {
            return await http.GetAsync<UserInfo>($"/{userId}");
        }

        public async Task<UserInfo[]> GetUsersAsync(string userIds)
        {
            return await http.GetAsync<UserInfo[]>($"/users/${userIds}");
        }

        public async Task<UserInfo> GetUserByEmailAsync(string email)
        {
            return await http.GetAsync<UserInfo>($"/by-email/{WebUtility.UrlEncode(email)}");
        }

        public async Task ResetPasswordAsync(string responseUrl, ResetPasswordRequest request)
        {
            await http.PutAsync($"/reset-password/{WebUtility.UrlEncode(responseUrl)}", request);
        }

        public async Task SignInAsync(SignInRequest request)
        {
            await http.PutAsync($"/sign-in", request);
        }

        public async Task SaveAsync(string responseUrl, SaveUserRequest request)
        {
            await http.PutAsync($"/{WebUtility.UrlEncode(responseUrl)}", request);
        }

        public async Task ConfirmAsync(ConfirmationTokenType type, Guid value)
        {
            await http.PutAsync($"/confirm/{type}/{value}");
        }

        public async Task ResendEmailConfirmationTokenAsync(string responseUrl)
        {
            await http.PostAsync($"/resend-email-confirmation/{responseUrl}");
        }

        public async Task SignUpAsync(string responseUrl, SignUpRequest request)
        {
            await http.PostAsync($"/{WebUtility.UrlEncode(responseUrl)}", request);
        }

        public async Task DeleteAsync()
        {
            await http.DeleteAsync("/");
        }
    }
}

File: App/icompanyportal/src/app/services/api-clients/users-api/users-client.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { User } from './user';
import { ResetPasswordRequest } from './reset-password-request';
import { SignInRequest } from './sign-in-request';
import { SignUpRequest } from './sign-up-request';
import { SaveUserRequest } from './save-user-request';
import { ConfirmationTokenType } from './confirmation-token-type';
import { USERS_API_ENDPOINT } from './users-api-endpoint';

@Injectable({
  providedIn: 'root'
})
export class UsersClient {
  constructor(private http: HttpClient, @Inject(USERS_API_ENDPOINT) private endpoint: string) {}

  doesUserExist(userId: number) {
    return this.http.get<boolean>(`${this.endpoint}/${userId}/exists`);
  }

  isEmailUnique(email: string) {
    return this.http.get<boolean>(`${this.endpoint}/${encodeURIComponent(email)}/is-unique`);
  }

  getSignedInUser() {
    return this.http.get<User>(`${this.endpoint}`);
  }

  getUser(userId: number) {
    return this.http.get<User>(`${this.endpoint}/${userId}`);
  }

  getUsers(userIds: string) {
    return this.http.get<User[]>(`${this.endpoint}/users/${userIds}`);
  }

  getUserByEmail(email: string) {
    return this.http.get<User>(`${this.endpoint}/by-email/${encodeURIComponent(email)}`);
  }

  resetPassword(responseUrl: string, request: ResetPasswordRequest) {
    return this.http.put<void>(`${this.endpoint}/reset-password/${encodeURIComponent(responseUrl)}`, request);
  }

  signIn(request: SignInRequest) {
    return this.http.put<void>(`${this.endpoint}/sign-in`, request);
  }

  signUp(responseUrl: string, request: SignUpRequest) {
    return this.http.post<void>(`${this.endpoint}/${encodeURIComponent(responseUrl)}`, request);
  }

  save(responseUrl: string, request: SaveUserRequest) {
    return this.http.put<void>(`${this.endpoint}/${encodeURIComponent(responseUrl)}`, request);
  }

  confirm(type: ConfirmationTokenType, value: string) {
    return this.http.put<void>(`${this.endpoint}/confirm/${type}/${value}`, undefined);
  }

  resendEmailConfirmationToken(responseUrl: string) {
    return this.http.post<void>(`${this.endpoint}/resend-email-confirmation/${encodeURIComponent(responseUrl)}`, undefined);
  }

  delete() {
    return this.http.delete<void>(this.endpoint);
  }
}
